ollama: packaging fixes

edef1c opened the pull request:
## Description of changes

llama.cpp was getting built twice: once in the buildGoModules fetch phase (immediately discarded), and then again in the actual build. Resolved by dropping `preBuild` from the module fetch derivation.

llama.cpp patches were applied through `patches` through a roundabout mechanism involving a second FOD for reasons that are utterly beyond me. I've replaced it with just applying the patches in `postPatch` so that we don't have to keep an extra list of hashes updated.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

edef1c committed:
ollama: don't build llama.cpp twice

buildGoModules runs preBuild hooks in the module download fixed-output
derivation, so llama.cpp got built and immediately discarded, to then
immediately be built again for the actual Go build.

abysssol reviewed (changes requested):
I'll merge after 86a062b41bdd4726403d6f36fdd7e50a86c5e747 is removed, since #326867 was recently merged. I can do the rebase if you're busy.

Everything else looks good.

edef1c committed:
ollama: apply llama.cpp patches in postPatch

This removes the absurd mechanism introduced to jam them into the
patches attribute, along with the pointless busywork of updating
the list of patches and their hashes on every release.

edef1c commented on pkgs/by-name/ol/ollama/package.nix:
```diff
@@ -52,30 +52,6 @@ let
-      stripLen = 1;
-      extraPrefix = "llm/llama.cpp/";
-    };
-
```
It *might* be worth putting the patch hook as `prePatch` or `postUnpack` rather than `postPatch` if we expect to apply patches of our own, or even in the `postFetch` of the src derivation, so that we can treat the loose patches as just an unfortunate distribution detail of the upstream source.

abysssol replied:
If you want to move the contents of `postPatch` to `postFetch`, I'd be totally fine with that; I can't think of any reason not to.

> Sticking it in postFetch would have the tiny bonus of not applying them in both the module fetch and the actual build

I don't think I understand: are you saying the patches are currently being applied twice? If so, please elaborate.

edef1c replied:
> I don't think I understand: are you saying the patches are currently being applied twice? If so, please elaborate.

There are two derivations: the Go module fetch one, which builds the vendor directory, and the actual build.
`patches` and `postPatch` are applied in both. The overhead is pretty negligible though.

Steinhagen reviewed (approved):
Everything looks and works as expected.
